core/utils/queue.h

```diff
@@ -0,0 +1,24 @@
+#ifndef BESS_UTILS_QUEUE_H
+#define BESS_UTILS_QUEUE_H
+
+namespace bess {
+namespace utils {
+
+// Takes Template argument T that is the type to be enqueued and dequeued.
+template <typename T>
+class Queue {
+ public:
+  virtual ~Queue(){};
+  virtual int Push(T*) = 0;
```

Comment:
I am a little bit concerned that the convention used here is quite different from general template usages.

For example, if someone wants to store "pointers" to objects, the template type T would be different between STL and bess::utils::Queue, which is quite confusing.
```
class Foo { ...} ;
Foo obj;

std::deque<Foo *> a;
a.push(&obj);

bess::utils::Queue<Foo> b;
b.push(&obj);
```

I understand that `llring` only takes pointer elements, but that is limitation of LockFreeQueue, not Queue. You can enforce only pointers to be used in LockFreeQueue, something like this (not tested):
```
#include <type_traits>

template<typename T>
class LockFreeQueue : Queue {
  static_assert(std::is_pointer<T>::value, "LockFreeQueue only support pointer types");
  ...
};
```